plugins/stylelint-suit-naming-pattern/index.js

```diff
@@ -0,0 +1,46 @@
+const stylelint = require("stylelint")
+
+const ruleName = "custom-rule/suit-naming-pattern"
+
+const messages = stylelint.utils.ruleMessages(ruleName, {
+  wrongCapitalization: (className) =>
+    `Component Class name "${className}" should follow the SUIT-style naming conventions with the ".spectrum" prefix.`,
+  wrongUtilityClassName: (className) =>  `Utility Class name "${className}" should follow the SUIT-style naming conventions with the ".u-camelCase" prefix.`,
+})
+const utilityClassPattern = /^\.u-[a-z]+[a-zA-Z0-9]+/;
+
+const suitClassPattern =
+/^spectrum((--|.)[a-z][a-z()]+)?((-[A-Z][a-zA-Z0-9:.()]+)([-]{1,2}[a-z][a-zA-Z-0-9():.]+)?(\s|,\s)?)?$/
+module.exports = stylelint.createPlugin(ruleName, () => {
+  return (root, result) => {
+    root.walkRules((rule) => {
+      rule.selectors.forEach((selector) => {
+        if (selector.includes(".spectrum")) {
+          const classNames = selector.split(/[.\s]+/).filter(className => className.startsWith('spectrum'));
+          const invalidCSS = classNames.every((className) => !suitClassPattern.test(className))
+            if (invalidCSS) {
+              stylelint.utils.report({
+                message: messages.wrongCapitalization(selector),
+                node: rule,
+                result,
+                ruleName,
+              })
+            }
+        } else if (selector.startsWith(".u")) {
+          // Utility pattern
+          if (!utilityClassPattern.test(selector)) {
+            stylelint.utils.report({
+              message: messages.wrongUtilityClassName(selector),
+              node: rule,
+              result,
+              ruleName,
+            })
+          }
+        }
+      })
+    })
+  }
```

Comment:
```suggestion
    root.walkRules(async (rule) => {
      await parser(selectors => {
        selectors.walkClasses(selectorNode => {
          const selector = selectorNode.value;

          if (selector.startsWith('spectrum') && !suitClassPattern.test(selector)) {
            stylelint.utils.report({
              message: messages.wrongCapitalization(selector),
              node: rule,
              result,
              ruleName,
            });
          } else if ((selector.startsWith("u-") || selector.startsWith("is-")) && !utilityClassPattern.test(selector)) {
            stylelint.utils.report({
              message: messages.wrongUtilityClassName(selector),
              node: rule,
              result,
              ruleName,
            });
          } else {
            stylelint.utils.report({
              message: messages.wrongClassName(selector),
              node: rule,
              result,
              ruleName,
            });
          }
        });
      }).process(rule.selector);
    })
  }
```

This approach allows you to target each individual class name without having to worry about valid pseudos or tag usage. First the postcss utility breaks the string apart into distinct classes and then iterates over them to check them against our rules specifically.

Reply:
Updated with postcss. I had to update the regex a bit to accomodate few patterns. Thanks for the idea.